pkg/eventemitter/eventemitter.go

```diff
@@ -0,0 +1,365 @@
+
+		e.checkEventHandlers(ctx, cloudEventSource, cloudEventSourceMutex)
+
+		select {
+		case <-tmr.C:
+			tmr.Stop()
+		case eventData := <-ch:
+			log.V(1).Info("Consuming events in queue.")
+			e.emitEventByHandler(eventData)
+		case <-ctx.Done():
+			tmr.Stop()
+			return
+		}
+	}
+}
+
+// checkEventHandlers will check each eventhandler active status
+func (e *EventEmitter) checkEventHandlers(ctx context.Context, cloudEventSource *eventingv1alpha1.CloudEventSource, cloudEventSourceMutex sync.Locker) {
+	cloudEventSourceMutex.Lock()
+	defer cloudEventSourceMutex.Unlock()
+	err := e.Client.Get(ctx, types.NamespacedName{Name: cloudEventSource.Name, Namespace: cloudEventSource.Namespace}, cloudEventSource)
+	if err != nil {
+		log.Error(err, "error getting cloudEventSource", "cloudEventSource", cloudEventSource)
+		return
+	}
+	keyprefix := cloudEventSource.GenerateIdentifier()
+	needUpdate := false
+	for k, v := range e.eventHandlersCache {
+		if strings.Contains(k, keyprefix) {
+			if v.GetActiveStatus() != cloudEventSource.Status.Conditions.GetActiveCondition().Status {
+				needUpdate = true
+				cloudEventSource.Status.Conditions.SetActiveCondition(metav1.ConditionFalse, v1alpha1.ScaledObjectConditionReadySucccesReason, v1alpha1.ScaledObjectConditionReadySuccessMessage)
+			}
+		}
+	}
+
+	if needUpdate {
+		cloudEventSourceStatus := cloudEventSource.Status.DeepCopy()
+
+		transform := func(runtimeObj client.Object, target interface{}) error {
+			status, ok := target.(*eventingv1alpha1.CloudEventSourceStatus)
+			if !ok {
+				return fmt.Errorf("transform target is not eventingv1alpha1.CloudEventSourceStatus type %v", target)
+			}
+			switch obj := runtimeObj.(type) {
+			case *eventingv1alpha1.CloudEventSource:
+				obj.Status = *status
+			default:
+			}
+			return nil
+		}
+
+		if err := kedastatus.TransformObject(ctx, e.Client, log, cloudEventSource, cloudEventSourceStatus, transform); err != nil {
+			log.Error(err, "Failed to update CloudEventSourceStatus")
+		}
+	}
+}
+
+// Emit is emitting event to both local kubernetes and custom CloudEventSource handler. After emit event to local kubernetes, event will inqueue and waitng for handler's consuming.
+func (e *EventEmitter) Emit(object runtime.Object, namesapce types.NamespacedName, eventtype, reason, message string) {
+	e.EventRecorder.Event(object, eventtype, reason, message)
+
+	if len(e.eventHandlersCache) == 0 {
+		return
+	}
+
+	name, _ := meta.NewAccessor().Name(object)
+	eventData := EventData{
+		namespace:  namesapce.Namespace,
+		objectName: name,
+		eventtype:  eventtype,
+		reason:     reason,
+		message:    message,
+		time:       time.Now().UTC(),
+	}
+	go e.inqueueEventData(eventData)
+}
+
+func (e *EventEmitter) inqueueEventData(eventData EventData) {
+	count := 0
+	for {
+		if count > MaxWaitingInqueueTIme {
+			log.Error(nil, "CloudEventSource channel is full and need to be check if handler can emit events")
+			return
+		}
+		select {
+		case ch <- eventData:
```

Comment:
Be careful here: `channel <-` is the syntax for sending to a channel, this `case` statement causes unpredictable results. Data can be written to `ch` `n`-number of times. The infinit loop is exited only after `count > MaxWaitingInqueueTIme` is true.

See example in [golang playground](https://go.dev/play/p/1NbvXKO9EsD)

Reply:
Each time there will be only one eventData pass through in the function and return directly once it was sent to the channel. So I think this will not be a problem. The purpose of this function is to keep trying to send one event data to the queue and skip if timeout occurs. Do you have any better implementation?

Reply:
Ahh I understand you now. Go By Example has a nice implementation of such use case: https://gobyexample.com/timeouts

```golang
func (e *EventEmitter) enqueueEventData(eventData EventData) {
	select {
	case ch <- eventData:
		log.V(1).Info("Event enqueued successfully.")
	case <-time.After(maxWaitingInqueueTime * time.Second):
		log.V(1).Info("Event cannot enqueue. Wait for next round.")
	}
}
```


And also the size of the buffer `10` is quite small. Can you increase it to `1024`?